src/misc-utils.ts

```diff
@@ -26,6 +46,148 @@ export async function getEntryStats(
   }
 }
 
+/**
+ * Asserts that the given object matches the shape outlined by the given
+ * Superstruct struct.
+ *
+ * @param object - The object to check.
+ * @param struct - The Superstruct object struct that you want to match against
+ * the object.
+ */
+export function assertJsonMatchesStruct<
+  Value extends Json,
+  Schema extends ObjectSchema,
+>(object: Json, struct: Struct<Value, Schema>): asserts object is Value {
+  try {
+    assert(object, struct);
+  } catch (error) {
+    if (error instanceof StructError) {
+      const failureMessages = error
+        .failures()
+        .map((failure) => {
+          return failure.message.endsWith('but received: undefined')
+            ? `Missing \`${String(failure.key)}\``
+            : `Invalid \`${String(failure.key)}\` (${failure.message})`;
+        })
+        .join('; ');
+
+      throw wrapError(
+        error,
+        `${failureMessages}.`,
+        INVALID_JSON_FILE_ERROR_CODE,
+      );
+    }
+    throw error;
+  }
+}
+
+/**
+ * Read the directory at the given path, and any directories underneath it, to
+ * arrive at all of the non-directories underneath that directory recursively.
+ *
+ * TODO: Move to @metamask/utils.
+ *
+ * @param directoryPath - The path to the directory.
+ * @param rootDirectoryPath - The path to another directory that will be used as
+ * the base for calculating relative paths for all entries. Defaults to
+ * `directoryPath`.
+ * @returns Objects representing all files underneath the directory.
+ * @throws An error with a stack trace if reading fails in any way.
+ */
+export async function readDirectoryRecursively(
+  directoryPath: string,
+  rootDirectoryPath = directoryPath,
+): Promise<DirectoryEntry[]> {
+  try {
+    const dirents = await fs.promises.readdir(directoryPath, {
+      withFileTypes: true,
+    });
+    const groupsOfChildEntries = await Promise.all(
+      dirents.map(async (dirent) => {
+        const fullPath = path.join(directoryPath, dirent.name);
+
+        if (dirent.isDirectory()) {
+          return await readDirectoryRecursively(fullPath, rootDirectoryPath);
+        }
+
+        const entry: DirectoryEntry = Object.assign({}, dirent, {
+          fullPath,
+          relativePath: path.relative(rootDirectoryPath, fullPath),
+        });
+        return [entry];
+      }),
+    );
+    return groupsOfChildEntries.flat();
+  } catch (error) {
+    throw wrapError(error, `Could not read directory '${directoryPath}'`);
+  }
+}
+
+/**
+ * Builds a new error object, linking it to the original error via the `cause`
+ * property if it is an Error.
+ *
+ * This is different from the `@metamask/utils` version because it allows for
```

Comment:
Nit: Have you considered using the `wrapError` function from utils, and overwriting the error code on the resulting error? Most of the logic is still the same, that would let us get rid of most of this complexity.

Reply:
Nice feature though, seems useful to add to the `utils` version of this function.

Reply:
> Nit: Have you considered using the `wrapError` function from utils, and overwriting the error code on the resulting error? Most of the logic is still the same, that would let us get rid of most of this complexity.

Good call — I've simplified it here: [52e9ce7](https://github.com/MetaMask/module-lint/pull/39/commits/52e9ce70d073c8f95cfe8933af068afcc3e89ccf)